pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -467,13 +483,22 @@ public SegmentMetadata getSegmentMetadata() {
     return physicalColumnNames;
   }
 
+  @Override
+  public Set<String> getColumnNamesForSelectStar() {
+    return Sets.union(getPhysicalColumnNames(), _newlyAddedColumnsFieldMap.keySet());
+  }
+
   @Override
   public ColumnDataSource getDataSource(String columnName) {
     FieldSpec fieldSpec = _schema.getFieldSpecFor(columnName);
-    if (fieldSpec.isVirtualColumn()) {
+    if ((fieldSpec == null && _newlyAddedColumnsFieldMap.containsKey(columnName)) || fieldSpec.isVirtualColumn()) {
```

Comment:
We should have a different column provider here that should be optimized better than the virtual column provider. The new column provider should not be built every time, but should stay in memory once built. What  may be updated is the number of rows. That will minimize additional allocation during query processing.

Add a new column provider factory (say, ConstantValueColumnProviderFactory) that has a map from the col name to the provider. The map is built lazily as columns are added to the query.

Reply:
In that case, this logic is best handled like:
```
if (fieldSpec != null) {
  if (fieldSpec.isVirtualColumn()) {
    // process virtual column from schema
  } else {
    // process real column from schema
  }
} else {
  // process newly added column using the new provider factory
}
```

Reply:
I don't see a strong reason to create another column provider factory just for this. I have updated the MutableSegmentImpl to cache the provider, and only initialize the provider or update row count when needed. This optimization will help other virtual columns ($segmentName and $hostName) too.